File: factory/cpp/factory.cpp
#include <iostream>
using namespace std;

//Products

class Button {
    public:
        virtual ~Button() {};
        virtual void render() = 0;
        virtual void onClick() = 0;
};

class LinuxButton: public Button {
    public:
        void render() {
            cout << "LinuxButton rendered" << endl;
        }
        void onClick() {
            cout << "LinuxButton onClick" << endl;
        }
};

class WindowsButton: public Button {
    public:
        void render() {
            cout << "WindowsButton rendered" << endl;
        }
        void onClick() {
            cout << "WindowsButton onClick" << endl;
        }
};

//Factories

class ButtonFactory {
    public:
        virtual ~ButtonFactory() {};
        virtual Button* createButton() = 0;
};

class LinuxButtonFactory: public ButtonFactory {
    public:
        Button* createButton() {
            return new LinuxButton();
        }
};

class WindowsButtonFactory: public ButtonFactory {
    public:
        Button* createButton() {
            return new WindowsButton();
        }
};

//Client code

class Dialog {
    public:
        Dialog(ButtonFactory &buttonFactory): buttonFactory(&buttonFactory), button(NULL) {};
        ~Dialog() {
            if (this->button != NULL)
                delete this->button;
        }

        string checkOS() {
            return "LINUX";
        }

        void render() {
            this->button = this->buttonFactory->createButton();
            this->button->render();
        }
    private:
        ButtonFactory *buttonFactory;
        Button* button;
};

int main(void) {
    WindowsButtonFactory factory;
    Dialog dialog(factory);
    dialog.render();
    return 0;
}

File: factory/typescript/factory.ts
// Products

abstract class Button {
    abstract render(): void;
}

class LinuxButton extends Button {
    public render(): void {
        console.log('LinuxButton');
    }
};

class WindowsButton extends Button {
    public render(): void {
        console.log('WindowsButton');
    }
};

//Factories

interface IButtonFactory {
    createButton(): Button;
};

class LinuxButtonFactory implements IButtonFactory {
    public createButton(): Button {
        return new LinuxButton();
    }
};

class WindowsButtonFactory implements IButtonFactory {
    public createButton(): Button {
        return new WindowsButton();
    }
};

// Client code

class Dialog {
    private factory: IButtonFactory;

    constructor(factory: IButtonFactory) {
        this.factory = factory;
    }

    public render(): void {
        const button = this.factory.createButton();
        button.render();
    }
};

const factory = new WindowsButtonFactory();
const dialog = new Dialog(factory);
dialog.render();
